gk/main.c

```diff
@@ -616,15 +773,35 @@ add_new_flow_from_policy(
 		return NULL;
 	}
 
-	/* Create a new flow entry. */
+	/*
+	 * We double the hash table size to alleviate memory pressure
+	 * when the table is full.
+	 */
 	ret = rte_hash_add_key_with_hash(
 		instance->ip_flow_hash_table,
  		&policy->flow, rss_hash_val);
-	if (ret < 0) {
+	if (ret == -EINVAL) {
 		RTE_LOG(ERR, HASH,
-			"The GK block failed to add new key to hash table in %s!\n",
+			"The GK block failed to add new key to hash table in %s with invalid parameters!\n",
 			__func__);
 		return NULL;
+	} else if (ret == -ENOSPC) {
+		RTE_LOG(WARNING, HASH,
+			"The GK block failed to add new key to hash table in %s due to lack of space!\n",
+			__func__);
+		ret = expand_flow_table(lcore_id, instance, gk_conf);
+		if (ret < 0)
+			return NULL;
+
+		ret = rte_hash_add_key_with_hash(
+			instance->ip_flow_hash_table,
+			&policy->flow, rss_hash_val);
+		if (ret < 0) {
+			RTE_LOG(ERR, HASH,
+				"The GK block failed to add new key to the expanded hash table in %s!\n",
+				__func__);
+			return NULL;
+		}
```

Comment:
Same comment as below -- can we make use of a `goto` here to avoid duplicating this code?